### PR TITLE
Fix IDE0009 false positive on local functions

### LINKED ISSUE #38043: IDE0009 false positive on local functions
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/695869/ide0009-false-positive-on-local-functions.html)._

---
[regression] [worked-in:16.3.0 Preview 2.0]
On some local functions IDE0009 is triggered e.g.:

```csharp
private void SomeFunction()
{
   object LocalFunction() => new object();
   this. DoSomethingElse(LocalFunction);
}
```

When invoking &quot;LocalFunction&quot; from whithin &quot;SomeFunction&quot;, IDE0009 is triggered which is not possible to be fixed.

---
#### Visual Studio Feedback System on 8/15/2019, 09:16 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 8/15/2019, 10:51 PM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by allisonchou**
Fixes #38043 and adds relevant test.

### CODE REVIEW

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -152,7 +152,8 @@ private bool IsStaticMemberOrTargetMethod(IOperation operation)
             switch (operation)
             {
                 case IMemberReferenceOperation memberReferenceOperation:
-                    return memberReferenceOperation.Member == null || memberReferenceOperation.Member.IsStatic;
+                    return memberReferenceOperation.Member == null || memberReferenceOperation.Member.IsStatic ||
+                        memberReferenceOperation.Member is IMethodSymbol method && method.MethodKind == MethodKind.LocalFunction;
                 case IInvocationOperation invocationOperation:
                     return invocationOperation.TargetMethod == null || invocationOperation.TargetMethod.IsStatic;
```
Comment (mavasani) [inline src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: Do we also need to handle the same case for the second switch section, i.e.

```
using System;
class C
{
    void Method()
    {
        object LocalFunction() => new object();
        |LocalFunction|]();
    }
    void Method2(Func<object> LocalFunction)
    {
    }
}
```

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -152,7 +152,8 @@ private bool IsStaticMemberOrTargetMethod(IOperation operation)
             switch (operation)
             {
```
Comment (mavasani) [inline src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: You may also want rename the method to indicate it returns true for all local functions, i.e. say `IsMemberOrTargetMethodStaticOrLocalFunction`

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -147,14 +147,16 @@ private void AnalyzeOperation(OperationAnalysisContext context, IOperation opera
...
 
-        private bool IsStaticMemberOrTargetMethod(IOperation operation)
+        private bool IsMemberOrTargetMethodStaticOrLocalFunction(IOperation operation)
         {
             switch (operation)
             {
                 case IMemberReferenceOperation memberReferenceOperation:
-                    return memberReferenceOperation.Member == null || memberReferenceOperation.Member.IsStatic;
+                    return memberReferenceOperation.Member == null || memberReferenceOperation.Member.IsStatic ||
+                        memberReferenceOperation.Member is IMethodSymbol memberMethod && memberMethod.MethodKind == MethodKind.LocalFunction;
```
Comment (RikkiGibson) [inline src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: consider using a pattern like `memberReferenceOperation.Member is IMethodSymbol { MethodKind: MethodKind.LocalFunction }`

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs (hunk 1 above)
Comment (allisonchou) [reply src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: Fixed!

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs (hunk 2 above)
Comment (allisonchou) [reply src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: Got it, just renamed.

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs (hunk 3 above)
Comment (allisonchou) [reply src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: Thank you for the tip! Updated.

**Comment by sharwell** [comment]
This is similar to how DotNetAnalyzers/StyleCopAnalyzers#3023 is resolved, and goes back to a change in the meaning of `IsStatic` that was implemented Roslyn 3.3. 👍

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -147,14 +147,16 @@ private void AnalyzeOperation(OperationAnalysisContext context, IOperation opera
...
             switch (operation)
             {
                 case IMemberReferenceOperation memberReferenceOperation:
-                    return memberReferenceOperation.Member == null || memberReferenceOperation.Member.IsStatic;
+                    return memberReferenceOperation.Member == null || memberReferenceOperation.Member.IsStatic ||
+                        memberReferenceOperation.Member is IMethodSymbol { MethodKind: MethodKind.LocalFunction };
                 case IInvocationOperation invocationOperation:
-                    return invocationOperation.TargetMethod == null || invocationOperation.TargetMethod.IsStatic;
+                    return invocationOperation.TargetMethod == null || invocationOperation.TargetMethod.IsStatic ||
+                        invocationOperation.TargetMethod is IMethodSymbol { MethodKind: MethodKind.LocalFunction };
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: can we extract a helper that boty of these codepaths should call?

File: src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -147,14 +147,16 @@ private void AnalyzeOperation(OperationAnalysisContext context, IOperation opera
             }
         }
 
-        private bool IsStaticMemberOrTargetMethod(IOperation operation)
+        private bool IsMemberOrTargetMethodStaticOrLocalFunction(IOperation operation)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/QualifyMemberAccess/AbstractQualifyMemberAccessDiagnosticAnalyzer.cs]: Feels like a better name would be `IsStaticMemberOrIsLocalFunction`.

**Pull request merged**